pkg/controller/core.oam.dev/v1alpha2/application/application_finalizer_test.go

```diff
@@ -216,99 +202,31 @@ var _ = Describe("Test application controller finalizer logic", func() {
 		}
 		Expect(k8sClient.Create(ctx, app)).Should(BeNil())
 		By("Create a cross workload trait app")
-		_, err := reconciler.Reconcile(ctrl.Request{NamespacedName: appKey})
-		Expect(err).Should(BeNil())
+		reconcileRetry(reconciler, ctrl.Request{NamespacedName: appKey})
 		checkApp := &v1beta1.Application{}
 		Expect(k8sClient.Get(ctx, appKey, checkApp)).Should(BeNil())
 		Expect(checkApp.Status.Phase).Should(Equal(common.ApplicationRunning))
-		Expect(len(checkApp.Finalizers)).Should(BeEquivalentTo(0))
+		Expect(len(checkApp.Finalizers)).Should(BeEquivalentTo(1))
 		rt := &v1beta1.ResourceTracker{}
-		Expect(k8sClient.Get(ctx, getTrackerKey(checkApp.Namespace, checkApp.Name), rt)).Should(BeNil())
-		_, err = reconciler.Reconcile(ctrl.Request{NamespacedName: appKey})
-		Expect(err).Should(BeNil())
+		Expect(k8sClient.Get(ctx, getTrackerKey(checkApp.Namespace, checkApp.Name, "v1"), rt)).Should(BeNil())
+		reconcileRetry(reconciler, ctrl.Request{NamespacedName: appKey})
 		checkApp = new(v1beta1.Application)
 		Expect(k8sClient.Get(ctx, appKey, checkApp)).Should(BeNil())
 		Expect(len(checkApp.Finalizers)).Should(BeEquivalentTo(1))
 		Expect(checkApp.Finalizers[0]).Should(BeEquivalentTo(resourceTrackerFinalizer))
-		Expect(checkApp.Status.ResourceTracker.UID).Should(BeEquivalentTo(rt.UID))
 		Expect(len(rt.Status.TrackedResources)).Should(BeEquivalentTo(2))
 		By("Update the app, set type to normal-worker")
 		checkApp.Spec.Components[0].Traits = nil
 		Expect(k8sClient.Update(ctx, checkApp)).Should(BeNil())
-		_, err = reconciler.Reconcile(ctrl.Request{NamespacedName: appKey})
-		Expect(err).Should(BeNil())
+		reconcileRetry(reconciler, ctrl.Request{NamespacedName: appKey})
 		rt = &v1beta1.ResourceTracker{}
 		checkApp = new(v1beta1.Application)
 		Expect(k8sClient.Get(ctx, appKey, checkApp)).Should(BeNil())
-		Expect(k8sClient.Get(ctx, getTrackerKey(checkApp.Namespace, checkApp.Name), rt)).Should(BeNil())
-		Expect(checkApp.Status.ResourceTracker.UID).Should(BeEquivalentTo(rt.UID))
+		Expect(k8sClient.Get(ctx, getTrackerKey(checkApp.Namespace, checkApp.Name, "v2"), rt)).Should(BeNil())
 		Expect(len(rt.Status.TrackedResources)).Should(BeEquivalentTo(1))
 		Expect(k8sClient.Delete(ctx, checkApp)).Should(BeNil())
-		_, err = reconciler.Reconcile(ctrl.Request{NamespacedName: appKey})
-		Expect(err).Should(BeNil())
-		Expect(k8sClient.Get(ctx, getTrackerKey(checkApp.Namespace, checkApp.Name), rt)).Should(util.NotFoundMatcher{})
-	})
-})
-
-var _ = Describe("Test finalizer related func", func() {
```

Comment:
Do we still make them back? Maybe use `PDescribe`?

Reply:
The function being tested in this case has been removed. So I remove the whole case. I will add new cases for finalizers logic.